feat: initial project setup

acoent opened the pull request:
Added controllers, DTOs, models, mappers, repositories, and services for bookings, items, users, and item requests. Implemented in-memory storage, validation, and exception handling for all modules. Introduced global exception handler and custom exceptions for error management. Updated pom.xml to include Lombok version property.

acoent committed:
feat: initial project setup

Added controllers, DTOs, models, mappers, repositories, and services for bookings, items, users, and item requests. Implemented in-memory storage, validation, and exception handling for all modules. Introduced global exception handler and custom exceptions for error management. Updated pom.xml to include Lombok version property.

acoent committed:
Refactor code style

vlolad commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDto.java:
```diff
@@ -1,7 +1,63 @@
+    public BookingDto() {
+    }
+
+    public Long getId() {
```
А как же ломбок....

vlolad commented on src/main/java/ru/practicum/shareit/booking/mapper/BookingMapper.java:
```diff
@@ -0,0 +1,31 @@
+import ru.practicum.shareit.booking.model.Booking;
+import ru.practicum.shareit.booking.model.BookingStatus;
+
+public class BookingMapper {
```
Можно добавить Mapstruct в работу, будет проще делать мапперы

vlolad commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -0,0 +1,63 @@
+        return status;
+    }
+
+    public void setStatus(BookingStatus status) {
```
А как же ломбок... (2)

Геттеры/Сеттеры и конструктор без параметров надо заменить на аннотации. И так для всех классов
`@Data` в классах моделях лучше не использовать, т.к. потом при добавлении JPA могут возникнуть ошибки (пока просто рекомендация, это будет дальше в уроках. На будущее, если интересно: https://habr.com/ru/companies/haulmont/articles/564682)

vlolad commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,100 @@
+
+    @Override
+    public BookingDto create(Long userId, BookingDto dto) {
+        if (userId == null || !userRepo.existsById(userId)) throw new NotFoundException("User not found: " + userId);
```
По код-стайлу if-else всегда пишем через фигурные скобки: https://www.magnumblog.space/java/translating-java-code-conventions#if_else_statements

И ещё лучше с отступами, сейчас сложно читать код

vlolad commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,100 @@
+        if (dto.getStart() == null || dto.getEnd() == null)
+            throw new BadRequestException("Start and end must be provided");
+        if (!dto.getStart().isBefore(dto.getEnd())) throw new BadRequestException("Start must be before end");
+        if (dto.getEnd().isBefore(LocalDateTime.now())) throw new BadRequestException("End must be in the future");
```
Валидацию надо вынести в контроллер. В идеале сделать через аннотации

vlolad commented on src/main/java/ru/practicum/shareit/item/repository/InMemoryItemRepository.java:
```diff
@@ -0,0 +1,38 @@
+
+    public List<Item> searchAvailableByText(String text) {
+        String q = text.toLowerCase();
+        return items.values().stream().filter(i -> Boolean.TRUE.equals(i.getAvailable())).filter(i -> (i.getName() != null && i.getName().toLowerCase().contains(q)) || (i.getDescription() != null && i.getDescription().toLowerCase().contains(q))).sorted(Comparator.comparing(Item::getId)).collect(Collectors.toList());
```
Разделить построчно условия, это нечитаемо

vlolad commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,134 @@
+    private final InMemoryItemRepository itemRepo;
+    private final InMemoryUserRepository userRepo;
+    private final InMemoryBookingRepository bookingRepo;
+    private final ConcurrentMap<Long, List<CommentDto>> comments = new ConcurrentHashMap<>();
```
Комментарии тоже должны храниться в репозитории

vlolad commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,134 @@
+        if (!userRepo.existsById(userId)) throw new NotFoundException("User not found: " + userId);
+        Item item = itemRepo.findById(itemId).orElseThrow(() -> new NotFoundException("Item not found: " + itemId));
+
+        boolean ok = bookingRepo.findByBookerIdAndItemId(userId, itemId).stream().anyMatch(b -> b.getStatus() == BookingStatus.APPROVED && (b.getEnd().isBefore(LocalDateTime.now()) || b.getEnd().isEqual(LocalDateTime.now())));
```
Нейминг переменных должен отражать то, для чего они нужны

vlolad commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,134 @@
+
+    private <T> List<T> paginate(List<T> list, int from, int size) {
+        if (from < 0) throw new BadRequestException("from must be >= 0");
+        if (size <= 0) throw new BadRequestException("size must be > 0");
```
Должно проверяться ещё на этапе контроллера

vlolad reviewed (changes requested):
Общая структура сделана, но много замечаний

1. Валидацию надо перенести в контроллер. Желательно сделать через аннотации с помощью валидатора хибернейта. Дальше будет только больше кода, это хоть как-то его сократит
2. Очень много ошибок по код-стайлу. Нет отступов, длинные строки с множеством условий и отсутствие фигурных скобок. На все моменты комментарии не оставлял, надо вычитать работу повторно и поправить это
3. Ломбок в проект добавлен, но не используется. По требованиям к работам - должен использоваться

Отдельный комментарий про многопоточку - вижу, что часто используются и потокобезопасные классы, но это лишнее. Не ошибка, просто это не нужно в рамках текущего проекта. А после добавления БД фреймворк сам уже будет (по большей части) контролировать транзакции, так что надо будет всё равно убирать
Это к тому, что на текущем спринте сильно углубляться в это не надо с:

acoent committed:
Add validation, mappers, and comment support

Introduced Jakarta Bean Validation annotations and custom validators for DTOs, added MapStruct mappers for entities, and refactored controllers and services to use validation and mapping. Implemented in-memory comment repository and model, added comment support to item endpoints, and improved exception handling with a unified error response. Replaced manual mapping and boilerplate with Lombok and MapStruct, and updated booking logic and status enum for consistency.

vlolad commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<BookingDto> create(@RequestHeader("X-Sharer-User-Id") Long userId, @RequestBody @Valid BookingDto dto) {
```
`"X-Sharer-User-Id"` используется часто и в нескольких контроллерах. Лучше всего сделать отдельный класс под глобальные константы проекта и перенести туда

vlolad commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,75 @@
+    public BookingDto create(Long userId, BookingDto dto) {
+        if (!userRepo.existsById(userId)) throw new NotFoundException("User not found: " + userId);
+        if (dto == null) throw new BadRequestException("Booking data required");
+        if (dto.getStart() == null || dto.getEnd() == null) throw new BadRequestException("Start and end required");
```
Это валидируется в контроллере аннотациями, надо убрать дублирующую проверку

vlolad commented on src/main/java/ru/practicum/shareit/booking/validation/StartBeforeEnd.java:
```diff
@@ -0,0 +1,18 @@
+@Retention(RetentionPolicy.RUNTIME)
+@Constraint(validatedBy = StartBeforeEndValidator.class)
+@Documented
+public @interface StartBeforeEnd {
```
Где эта аннотация используется? Надо либо использовать, либо убрать

vlolad commented on src/main/java/ru/practicum/shareit/item/validation/ItemAvailableValidator.java:
```diff
@@ -0,0 +1,26 @@
+import ru.practicum.shareit.item.repository.InMemoryItemRepository;
+
+@Component
+public class ItemAvailableValidator implements ConstraintValidator<ItemAvailable, Long> {
```
В Валидаторе запросы через репозиторий делать не надо, это уже задача сервиса. Надо убрать

vlolad commented on src/main/java/ru/practicum/shareit/item/validation/ItemExistsValidator.java:
```diff
@@ -0,0 +1,24 @@
+import ru.practicum.shareit.item.repository.InMemoryItemRepository;
+
+@Component
+public class ItemExistsValidator implements ConstraintValidator<ItemExists, Long> {
```
Аналогично, запросы к базе через валидаторы делать не надо

vlolad commented on src/main/java/ru/practicum/shareit/request/service/ItemRequestServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+    @Override
+    public ItemRequestDto create(Long userId, ItemRequestDto dto) {
+        if (!userRepo.existsById(userId)) throw new NotFoundException("User not found: " + userId);
+        if (dto == null || dto.getDescription() == null || dto.getDescription().isBlank()) {
```
Проверка в контроллере через аннотации, тут дублируется

vlolad commented on src/main/java/ru/practicum/shareit/request/service/ItemRequestServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+    public List<ItemRequestDto> getAll(Long userId, int from, int size) {
+        if (!userRepo.existsById(userId)) throw new NotFoundException("User not found: " + userId);
+        if (from < 0) throw new BadRequestException("from must be >= 0");
+        if (size <= 0) throw new BadRequestException("size must be > 0");
```
Значения по умолчанию также должен обеспечивать контроллер

vlolad commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,45 @@
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Size;
+
+public class UserDto {
```
Аннотации ломбока пропущены

vlolad commented on src/main/java/ru/practicum/shareit/user/validation/UniqueEmailValidator.java:
```diff
@@ -0,0 +1,26 @@
+import ru.practicum.shareit.user.repository.UserRepository;
+
+@Component
+public class UniqueEmailValidator implements ConstraintValidator<UniqueEmail, String> {
```
Аналогично, в валидаторах запросы в базу не делаем. Надо убрать

vlolad commented on src/main/java/ru/practicum/shareit/validation/UserExistsValidator.java:
```diff
@@ -0,0 +1,24 @@
+import ru.practicum.shareit.user.repository.InMemoryUserRepository;
+
+@Component
+public class UserExistsValidator implements ConstraintValidator<UserExists, Long> {
```
Аналогично, в валидаторах запросы в базу не делаем. Надо убрать

vlolad reviewed (changes requested):
Добавлено много новых валидаторов, но ни один не используется. Большую часть надо убрать (в них спрятана логика по работе с запросами из базы), оставшуюся - либо как-то использовать, либо тоже убрать.
Ну и убрать дублирующие проверки из сервисов, так как валидиация через аннотации добавлена верно

Всё в комментариях

acoent committed:
refactor: remove DB logic from validators, consolidate validation and header

acoent committed:
style: satisfy Checkstyle by adding no-op comment to HeaderConstants()

acoent committed:
style: satisfy Checkstyle by adding no-op comment to HeaderConstants()

acoent committed:
style: satisfy Checkstyle
